lib/schema.js

```diff
@@ -428,21 +464,41 @@ class Metaschema {
           value = def(value);
           obj[field] = value;
         } else {
-          value = this.createInstance(def, value);
+          if (def.category) value = this.createInstance(def.category, value);
+          if (def.domain) value = this.createDomainInstance(def.domain, value);
           if (value !== null) obj[field] = value;
           else return null;
         }
       }
-      for (const prop of required) {
-        if (obj[prop] === null || obj[prop] === undefined) {
-          return null;
+      for (const prop of properties) {
+        const def = definition[prop];
+        if (!obj.hasOwnProperty(prop)) {
+          if (def.default) {
+            let value = null;
+            value = this.createDomainInstance(def.domain, def.default);
+            if (value === null) return null;
+            obj[prop] = value;
+          } else if (def.required) {
+            return null;
+          }
```

Comment:
Also, I'd merge those branches:
```js
if (!obj.hasOwnProperty(prop)) {
  if (def.default) {
    obj[prop] = this.createDomainInstance(def.domain, def.default);
  }
  if (obj[prop] == null && def.required) return null;
}
```

Reply:
We should return `null` even if `obj[prop] === null` and property is not required.